pytensor/tensor/rewriting/linalg.py

```diff
@@ -401,30 +404,80 @@ def _find_diag_from_eye_mul(potential_mul_input):
     eye_input = [
         mul_input
         for mul_input in inputs_to_mul
-        if mul_input.owner and isinstance(mul_input.owner.op, Eye)
+        if mul_input.owner
+        and (
+            isinstance(mul_input.owner.op, Eye)
+            or
+            # This whole condition checks if there is an Eye hiding inside a DimShuffle.
+            # This arises from batched elementwise multiplication between a tensor and an eye, e.g.:
+            # tensor(shape=(None, 3, 3) * eye(3). This is still potentially valid for diag rewrites.
+            (
+                isinstance(mul_input.owner.op, DimShuffle)
```

Comment:
Right, we only want DimShuffles that work out to `ExpandDims`. But that's non-trivial check, judging from the `__str__` method. What do you think of:

```python
                isinstance(mul_input.owner.op, DimShuffle)
                and mul_input.owner.op.__str__().startswith('ExpandDims')
                and mul_input.owner.inputs[0].owner is not None
                and isinstance(mul_input.owner.inputs[0].owner.op, Eye)
```

As a hacky-shortcut?

Reply:
No, we should make it a property of the DS Op, which can be done at init time.

```python
input_ndim = len(input.broadcastable)
self.is_left_expand_dims = (
  "x" in order
  and (
    input_ndim == 0
    or order[-input_ndim:] == tuple(range(input_ndim)
  )
```

Reply:
I can do that, but why don't these three different Ops (`Transpose`, `ExpandDims` and `DimShuffle`) just exist as wrappers we can check for? It was very confusing to me when I searched for the ExpandDims Op only to find it didn't exist

Reply:
Those are not Ops, it's just different labels we use in the string representation for more readability. It's all Dimshuffle

Reply:
Also this doesn't seem to catch the case we're testing for? Here's what I'm seeing, I guess it's expanding right?

```python
 import pytensor.tensor as pt
x = pt.tensor('x', shape=(None, 3, 3))
eye = pt.eye(3)
(eye * x).owner.inputs[0].owner.op.new_order

('x', 0, 1)
```

Reply:
> Those are not Ops, it's just different labels we use in the string representation for more readability. It's all Dimshuffle.

That's my point, since they aren't `Op` its confusing that they show up in a dprint as if they were

Reply:
Feel free to revert but good luck explaining people what a DimShuffle is

Reply:
pprint can also rename ops, that's why I took the liberty when trying to make dprint more readable. I'm open to reverting it but don't love it

Reply:
No I'm not going to revert. My suggestion was that the code that we currently call the `DimShuffle` Op should be a dispatcher to three (nearly) identical Ops: `Transpose`, `ExpandDims`, and `DimShuffle`.

I'm letting it go though, the scope creep is real. Let know what you think of where I landed on the `is_xxx_expand_dims` flags.

Reply:
The 3 Ops could make sense but that's a huge structural change to PyTensor. 

Also ordering would be a PITA, what if you have expand_dims after transpose followed by a squeeze (so functionally just a transpose) or the other way around? Now your rewrites need to worry about that. What would be a canonical order?

I think conceptually it's not a terrible idea, but it's something that doesn't exist in other libraries